src/units/abilities.cpp

```diff
@@ -1171,14 +1171,197 @@ unit_ability_list attack_type::get_special_ability(const std::string& ability) c
 	return abil_list;
 }
 
-bool attack_type::bool_ability(const std::string& ability) const
+	/**
+	 * Gets the children of parent (which should be the abilities for an
+	 * attack_type) and places the ones whose tag or id= matches @a id into
+	 * @a tag_result and @a id_result.
+	 * @param tag_result receive the children whose tag matches @a id
+	 * @param id_result receive the children whose id matches @a id
+	 * @param parent the tags whose contain children (abilities here)
+	 * @param id tag or id of child tested
+	 * @param special_id if true, children check by id
+	 * @param special_tags if true, children check by tags
+	 */
+static void get_ability_children(std::vector<special_match>& tag_result,
+	                           std::vector<special_match>& id_result,
+	                           const config& parent, const std::string& id,
+	                           bool special_id=true, bool special_tags=true) {
+	if(special_id && special_tags){
+		get_special_children(tag_result, id_result, parent, id);
+	} else if(special_id && !special_tags){
+		get_special_children_id(id_result, parent, id);
+	} else if(!special_id && special_tags){
+		get_special_children_tags(tag_result, parent, id);
+	}
+}
+
+bool unit::get_self_ability_bool(const config& special, const std::string& tag_name, const map_location& loc) const
+{
+	return (ability_active(tag_name, special, loc) && ability_affects_self(tag_name, special, loc));
+}
+
+bool unit::get_adj_ability_bool(const config& special, const std::string& tag_name, int dir, const map_location& loc, const unit& from) const
+{
+	const auto adjacent = get_adjacent_tiles(loc);
+	return (affects_side(special, side(), from.side()) && from.ability_active(tag_name, special, adjacent[dir]) && ability_affects_adjacent(tag_name,  special, dir, loc, from));
+}
+
+bool unit::get_self_ability_bool_weapon(const config& special, const std::string& tag_name, const map_location& loc, const_attack_ptr weapon, const_attack_ptr opp_weapon) const
+{
+	return (get_self_ability_bool(special, tag_name, loc) && ability_affects_weapon(special, weapon, false) && ability_affects_weapon(special, opp_weapon, true));
```

Comment:
Why does this need to check both `weapon` and `opp_weapon` for the special?

Reply:
this sub function is used for cheking [leadership] in special_id/type_active or this ability can't checke opponent but student and teacher only, and with [filter_weapon] and [filter_second_weapon], the weapon of student and opponent, then yes it need check weapon and opp_weapon for checking acitivtie of [leadership] abilities (filter diffrent of weapon special and abilities used like weapon

Reply:
I'm not sure I understand.  Wouldn't most cases be that the unit attacking and the unit defending don't have the same ability affecting their weapons?

Reply:
What I'm trying to say is that since leadership can filter the weapons of the unit under leadership as well as those of the adversary, special_id / type_active must be able to differentiate between active leadership and one that does not. would not be because one of the filtered weapons does not match like for an special active or not with [filter_student] or [filter_opponent]

Reply:
Alright, so for `unit::get_self_ability_bool_weapon`, it only gets called when the tag being checked is [leadership] and when `leader_bool` is set to true, which as of right now only happens in the `if(self_)` part of `attack_type::get_special_ability_bool`.  So the [leadership] check would only happen for the attacker.

Then for the `unit::get_self_ability_bool_weapon` method itself, it checks:
* `get_self_ability_bool(special, tag_name, loc)` - does the unit have this ability active right now and is the unit affected by it?
* `ability_affects_weapon(special, weapon, false)` - does this ability affect the attacker's weapon (by checking [filter_weapon])?
* `ability_affects_weapon(special, opp_weapon, true)` - does this ability affect the defender's weapon (by checking [filter_second_weapon])?

Is that correct?

Reply:
> 
> 
> Alright, so for `unit::get_self_ability_bool_weapon`, it only gets called when the tag being checked is [leadership] and when `leader_bool` is set to true, which as of right now only happens in the `if(self_)` part of `attack_type::get_special_ability_bool`. So the [leadership] check would only happen for the primary unit.
> 
> Then for the `unit::get_self_ability_bool_weapon` method itself, it checks:
> 
>     * `get_self_ability_bool(special, tag_name, loc)` - does the unit have this ability active right now and is the unit affected by it?
> 
>     * `ability_affects_weapon(special, weapon, false)` - does this ability affect the primary unit's weapon (by checking [filter_weapon])?
> 
>     * `ability_affects_weapon(special, opp_weapon, true)` - does this ability affect the secondary unit's weapon (by checking [filter_second_weapon])?
> 
> 
> Is that correct?

1. yes
2.  yes
3. yes

Reply:
Alright, so, what would be an example of a situation where the primary unit's leadership ability affects both the primary unit's weapon and the secondary unit's weapon?

Reply:
i make error for 3, leadership don't affect secondary weapon, but is conditonnal by this presence, but this the always the unit student who is affected by leadership. the first example is a magical leadership who work for magical weapon only, (use [filter_weapon]special_id_active=magical),
the second example is leadership who don't work if opponent use fire cold or weapon who afect the teacher(use [filter_second_weapon][not]type=cold). the ability_affect_weapon name can wrong the reader and all excuses for that.

Reply:
So given the following definitions:
* Teacher (unit with leadership ability)
* Primary unit (the unit attacking)
* Secondary unit (the unit defending)

The first example is of a teacher that has a leadership ability that only gives its bonuses to the primary unit's attacks that have the Magical weapon special.

The second example is of a teacher that has a leadership ability that only gives its bonuses to the primary unit's attacks that have the Magical weapon special, and only if the secondary unit isn't using an attack that has the Cold or Fire damage type.

Is that correct?

Reply:
Correct, but i didn't said what [filter_weapon]special_id_active=magical was implemented in second example

Reply:
Alright.  I think it'd be good for there to be new, separate WML unit tests that check the correctness of different cases where this can be used.

Reply:
you must stoping to use attacker and defender for define self/student and opponent in weapon special, attacker is the unit who begin the fight and apply_to=attacker aplied to unit who begin fight, student or opponent.
primary unit:(the unit student of Teacher, in leadership weapon filter by [filter_weapon])
secondary unit (the unit opponent to student weapon filtered by [filter_second_weapon] in leadership abilities)
with leadership ever active in offense and defense, using attacker and defender is confusing.

Reply:
test added for abilities affect_adjacent with or without [leadership] abilities